### ISSUE TITLE
Replace All operation can be extremely slow

### ISSUE THREAD

**Issue by rjsargent** [notepad-plus-plus/notepad-plus-plus#9318]
### Description of the Issue
Find and Replace All takes a long time

I have a 212KB file comprising ~5,500 lines. Changing ";class" to " class" (i.e. replace the semicolon with a space) took 28 minutes. 4,670 occurrences were replaced, ~3 occurrences per second.

Version: 7.9.0, 32-bit, Windows

### Steps to Reproduce the Issue
1. Create a file of the approximate size with 5,500 lines, 4670 of which end with ;class (see example below)
2. Do a Replace All, Normal, Match Case
3. Wait

### Expected Behavior
200KB should only take a few seconds to traverse

### Actual Behavior
28 minutes with Notepad++ showing as Not Responding

### Debug Information
Notepad++ v7.9   (32-bit)
Build time : Sep 22 2020 - 03:24:22
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2004
OS Build : 19041.685
Current ANSI codepage : 1252
Plugins : ComparePlugin.dll mimeTools.dll NppConverter.dll NppExport.dll NppXmlTreeviewPlugin.dll XMLTools.dll 

### Example (excerpt) of the file in question

Instances TotBytes AvgBytes Class
--------- -------- -------- -----
525781 10515620 20 Point
304641 6092820 20 Association
304155 12573948 41 Array
275502 11802092 43 OrderedCollection
209824 4196480 20 Rectangle
177989 4271736 24 MessageChannel
141020 3384480 24 BlockClosure
129200 8347023 65 ByteArray
...
1 48 48 Tools.FileTools.XMLSourceWalkerState;class
1 48 48 Tools.FileTools.XMLSourceWalkerTopState;class
1 52 52 Tools.FileTools.XMLTreeViewer;class
1 48 48 YellowButtonPressedEvent;class
1 48 48 YellowButtonReleasedEvent;class
1 52 52 GemStone.Gbs.Exceptions.ZeroDivide;class
1 52 52 ZeroDivide;class

3828208 157412563 41.1 *** All ***

**Comment by sasumner** [comment]
I must agree that is a ridiculous amount of time.  I tried generating some data of similar size and got nowhere in the annoying time range for any similar replacement.  Any chance you can upload your exact file so that it can be tried?  Otherwise I can't see this issue going anywhere...

**Comment by rjsargent** [comment]
I hope this will allow you to replicate my experience. This file is about 300KB, which is still not huge. But there are a lot of replacements going on. 4,670 lines out of 5,445 getting changed.

[InstanceCounts_Example_2020-12-29-16.45.49.848.txt](https://github.com/notepad-plus-plus/notepad-plus-plus/files/5752638/InstanceCounts_Example_2020-12-29-16.45.49.848.txt)

2012 Dell Latitude 6420 Intel Core i7-2640M @ 2.80 GHz 2 cores/4 logical processors (Windows Task Manager reports speed as 3.43 GHz)

28 minutes, 30 seconds using the replace conditions in the Sample attachment.
The Performance attachment shows that Notepad++ is not using a lot of memory.

![Notepad++_Sample](https://user-images.githubusercontent.com/4140292/103323625-a42c0f00-49f8-11eb-8d9d-595cd2420938.png)
![Notepad++_Performance](https://user-images.githubusercontent.com/4140292/103323630-a726ff80-49f8-11eb-8807-abf60bb34fd0.png)

FWIW, I suspect the Replace All algorithm is O(n^2) rather than something like a "double buffer" streaming algorithm.
(And one can stream in place if the replacement is equal to or smaller than the search string.)

p.s. I really like this editor!!

**Comment by sasumner** [comment]
>InstanceCounts_Example_2020-12-29-16.45.49.848.txt

When doing the replacement operation shown in the screenshot, my results were `4688 occurrences were replaced...` and it was hard to measure the time, but I want to say it took maybe half a second; certainly under one second.  This was with Notepad++ 7.9.1 64-bit.

Just to compare apples with apples I tried the same on 7.9 32-bit ... and got a similar half-a-second result.

These (below) were the only plugins I had installed; maybe your extra plugins make some sort of difference?
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

**Comment by dail8859** [comment]
I too was able to replace 4,670 instances in a fraction of a second.

> @sasumner
> When doing the replacement operation shown in the screenshot, my results were 4688 occurrences were replaced...

Make sure to search for ` class` (it includes a space in front of it).


@rjsargent

I suspect you tried some of these but here are some suggestions:

- Make sure nothing else is reading/writing to the file. Try moving the file to some other place (e.g. Desktop) and retry the replace.
- Try another editor perhaps. If you really want to try to debug this I'd suggest using [Scite](https://www.scintilla.org/SciTEDownload.html) since it uses the same internal library that would be doing the low-level find/replace.
- Temporarily disable plugins. You can run it from the command line with an extra parameter `C:\Program Files (x86)\Notepad++\notepad++.exe -noPlugin`
- Use some sort of process debugger. Think Windows has Process Monitor? But it can show low level details if its waiting on file IO, memory locks, etc.


---

For reference this is my debug info:

```
Notepad++ v7.9.1   (64-bit)
Build time : Nov  2 2020 - 01:07:46
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1809
OS Build : 17763.1637
Current ANSI codepage : 1252
Plugins : BetterMultiSelection.dll ComparePlugin.dll ElasticTabstops.dll LuaScript.dll mimeTools.dll NppConverter.dll NppExport.dll NppMarkdownPanel.dll SurroundSelection.dll XMLTools.dll 
```

**Comment by rjsargent** [comment]
Well, that is interesting. It does seem to be a plug-in problem!

The plug-ins I have installed are:
Compare, v2
Mime tools, v2.5
Npp Converter, v4.2.1
Npp Xml Treeview, v2
NppExport, v0.2.9
XML Tools, v3.0.3.3

I had assumed that, like Compare, the plug-ins don't have any effect unless activated from the plug-ins menu. Obviously, that's wrong.

I'll delete them one by one and see which one it is.

**Comment by rjsargent** [comment]
Well, that was quick. It was the first one I tried: XML Tools, v3.0.3.3

I wonder what the heck it was doing to so badly mess up the engine!

Thanks for the suggestions!

**Comment by rjsargent** [comment]
Well, I reloaded XML Tools, v3.0.4.3 and everything works fine. I can't believe a .. fix would have that impact.

There we have it, from release notes for 3.0.3.4:
Fix performance issue
Previous version may slow down NPP, especially during "Replace All" action.

------------------
Closing the issue.

**Comment by felix-22** [comment]
Related / duplicate of:  https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8351
